Add MassBlockUpdate interface.  Adds BUKKIT-4114.

desht opened the pull request:
### The Issue:

Bukkit's existing methods for block updates (Block.setType\* and Block.setData) are slow for large number of updates due to the following reasons: 1) A block update packet is sent for every change, and 2) Lighting is recalculated for every change.  This means a lot of server-side CPU load and network traffic as multiple packets are sent to every nearby player.
### Justification for this PR:

This PR creates an API to specify a mass block update interface, the implementation of which will allow direct block changes to be made without any packet sending until all changes have been made.  At that point, affected chunks are marked for refresh and will be sent by the server when it's ready.  In addition, the API allows for control for if and when lighting recalculations are carried out.
### PR Breakdown:

Adds MassBlockUpdate interface, plus method in World interface to create a MassBlockUpdate object for that world.

The interface allows for blocks to be set with setTypeId() and setTypeIdAndData() methods, for the changes to sent to interested clients with notifyClients(), and specifies a few other methods primarily for controlling how lighting will be calculated.
### Testing Results and Materials:

See: https://github.com/desht/buster2 which includes some test data in the README for the project.  Built version at https://dl.dropboxusercontent.com/u/12467600/buster2-1.0.jar

See also: http://dev.bukkit.org/server-mods/ChessCraft which has been successfully using equivalent methods (implemented via direct NMS access in the plugin) for well over a year.
### Relevant PRs:

CraftBukkit: https://github.com/Bukkit/CraftBukkit/pull/1150
### JIRA Tickets:

BUKKIT-4114 - https://bukkit.atlassian.net/browse/BUKKIT-4114

desht committed:
Add MassBlockUpdate interface.  Adds BUKKIT-4114.
Adds a MassBlockUpdate interface and associated method in Server to
get a MassBlockUpdate object for the World.  MassBlockUpdate specifies
an efficient mechanism to perform multiple block update operations.

Wolvereness commented on src/main/java/org/bukkit/block/MassBlockUpdate.java:
```diff
@@ -0,0 +1,140 @@
+         * that no lighting properties are being changed, or if your
+         * plugin will handle relighting itself.
+         */
+        NEVER,
```
`NEVER` should _never_ be in this type of implementation. Really, I don't see the point in a relighting _strategy_ at all. Mark the chunk as dirty, finish changing blocks, and immediately do a full relight all dirty chunks.

desht replied:
The point of relighting strategies is to give plugins some flexibility over how certain calculations need to be made.  I guess NEVER could be dropped, but there are certainly valid reasons why a plugin might want to choose IMMEDIATE/DEFERRED/STATIC options.  IMMEDIATE is only a little faster than plain old Bukkit (at least where block lighting properties have changed), and will certainly result in server lag for very large changes.  STATIC could be very useful to plugins who want to maintain a fixed lighting level over a region in the world, e.g. a game board.  DEFERRED is the best option for really huge changes where "natural" lighting is desired.

Wolvereness replied:
If it doesn't accomplish the task (`IMMEDIATE` as you said) it shouldn't be in a `MassBlockUpdate` class. As far as `STATIC` goes, read my comment below.

desht replied:
There's still good reason for using `IMMEDIATE` - 1) sending chunk updates is still better than sending an individual block update for every updated block, and 2) only blocks whose lighting properties have changed need to be re-lit (e.g. changing dirt to stone does not need a relighting calculation, because neither block emits light, and neither block blocks light)

EdGruberman replied:
This whole relighting strategy approach seems to require plugins to learn so much about the implementation that the API becomes pointless.  The implementation should manage lighting automatically in the best way appropriate.

Additionally, choices that allow the server and client to get out of sync on lighting seem counter-intuitive without some type of dedicated lighting API to manage it appropriately.

What are the use cases for needing to choose what type of lighting to perform and why can't the implementation be trusted to do what is appropriate instead?

riking replied:
> @desht ...neither block emits light, and neither block blocks light

Shouldn't that be **both** blocks block light?
What you typed would be applicable to changing a glass pane to an iron bar. Actually, none of those blocks are partially blocking (e.g. water), so maybe that's what you meant.

I think that only IMMEDIATE and DEFERRED should be available to plugins. By the way, are you neglecting skylight?

desht replied:
OK, use cases:
1) IMMEDIATE - normal case, where you want blocks re-lit as they're placed.  Heaviest on CPU, will cause lag for large changes.
2) NEVER - maximum performance.  But, as @Wolvereness and yourself pointed out, things can just look "wrong" (although the client seems to do a pretty good job at compensating for that when you get close to updated blocks, I guess that's an implementation detail that can't be relied on).
3) STATIC - for "flood-lighting" areas, such as game boards (my ChessCraft plugin is part of the reason I'm working on this).  Where you want to keep a given area at a given uniform light level.
4) DEFERRED - for very large changes, where you want to smooth out the load of re-lighting.  Lighting is done over several ticks.

So I admit NEVER may have to be ditched (fast but wrong isn't good enough), STATIC seems pretty useful to me, and IMMEDIATE/DEFERRED could potentially be merged, but the following need to be considered:
- DEFERRED could be the default approach here, but a good threshold value needs to be determined for the maximum time per tick that a server can spend doing re-lighting work.  Whether or not this is tunable by plugins is up for discussion.
- IMMEDIATE could be dropped entirely and made simply a special one-tick case of DEFERRED.

That would then boil down to a choice between static lighting and natural lighting (with a possibly tunable "aggressiveness" factor).

Regarding a dedicated lighting API, we're extremely limited in what we can do here.  The current protocol basically allows lighting to be updated as part of a chunk update block; there's no way of sending lighting updates for individual blocks.  So any lighting API will be limited to specifying the desired light level for one or more blocks and then sending chunk updates for the chunks that contain those blocks.  This is not something that should be encouraged for general dynamic lighting updates (e.g. a torchlight plugin where light follows the player), and @Wolvereness made the valid point below that this API could potentially be abused to do that without needing direct NMS access.  I'm not sure if there's any easy way of preventing that.

desht replied:
@riking Yeah, typo on my part there, sorry.  For dirt->stone, I meant that neither block emits light, and both blocks block light in the same way.  Therefore, no relighting calculation is necessary if one turns to the other, which was my original point.  For stone->glass, recalculation would be needed, because light-blocking has changed, and similarly for glass->water (although both transmit light, they transmit a different proportion of it).

Wolvereness commented on src/main/java/org/bukkit/block/MassBlockUpdate.java:
```diff
@@ -0,0 +1,140 @@
+ * block updates by consolidating the operations that would make
+ * multiple standard block updates very slow.
+ */
+public interface MassBlockUpdate {
```
This class really doesn't seem versatile enough. Sure, it's written like a queue of changes, but couldn't that be done using java API? Unless this is designed as a low-level block change interface (which absolutely reeks).

Personally, I'd be more inclined to take 'structure' (read: diff) based writing, to allow CraftBukkit to handle the internals of how to efficiently apply it; it's not particularly safe to assume this type of implementation can handle robust changes to the minecraft internals work.

desht replied:
Yeah, I appreciate this is skirting dangerously close to exposing implementation details (in particular the fact that updating a block id/data, recalculating lighting, and notifying clients are three different operations which Bukkit currently conflates into one method call).

I think I see where you're going with the "structure-based writing" idea.  One thing that occurred to me was a BlockUpdateProvider interface which plugins could implement (with perhaps a couple of stock implementations provided too).  This would be an iterator-like class which provided on-demand the next block update - basically a location plus block ID and maybe block data. MassBlockUpdate could then have a setBlockProvider() method, and a runBlockUpdate() method - runBlockUpdate() would simply consume block updates from the provider, make the changes, and when finished send the necessary chunk updates to clients (relighting is still a thorny issue, though).  This could be run over more than one tick in the interest of minimising server lag.  Thoughts?

(Stock implementations of BlockUpdateProvider could be one that pulls updates from a 1-d and/or 3-d array, and one that copies existing blocks from elsewhere).

Wolvereness replied:
Being completely candid, I don't want _any_ new interfaces to be implemented by plugins. I mean that more in the sense of "interface defines how it works, implement it it correctly please, and also, adding methods is breaking" and less in the "we have one method that we need you to implement (like commands) so we can call your code in this special case".

desht replied:
OK.  I'm not really sure how else this can be structured; the fundamental problem I want to address - that Bukkit block updates are currently slow because update/lighting/notification are conflated - is at its heart an implementation detail.  I'm completely open to suggestions on how a more robust interface would look.  Currently, it's "update some blocks; tell clients about the changes" - how else can it be structured?

Wolvereness commented on src/main/java/org/bukkit/block/MassBlockUpdate.java:
```diff
@@ -0,0 +1,140 @@
+     * @return whether the block was actually changed
+     * @throws IllegalStateException if this method is called after {@link #notifyClients()} is called
+     */
+    public boolean setTypeId(int x, int y, int z, int materialId);
```
Not having the `Material` overloads makes me VERY sad, especially when we are considering 4 ints. Locations are mutable, I'd have no issues forcing a plugin to have at least one location object to pass during calls.

desht replied:
I was paralleling the Block methods here, but I'm completely open to more overloads.  But which `Material` overloads are you referring to, specifically?

Wolvereness replied:
Why not a `set(Location, Material)` is what I mean.

desht replied:
Sure, that can easily be added.  Equivalent methods for all the existing block methods would be ideal, I guess.  Principle of least surprise and all that.

Wolvereness commented on src/main/java/org/bukkit/block/MassBlockUpdate.java:
```diff
@@ -0,0 +1,140 @@
+     *
+     * @return the number of blocks that still need re-lighting, or 0 if the relighting strategy is not RelightingStrategy.DEFERRED
+     */
+    public int getBlocksToRelight();
```
This method reeks of bad design. Like, REALLY bad design, as if the internal implementation was extremely specific. What is the true relevance to having this sort of method? Surely CB doesn't calculate every possible spot that a block would need updating because the light pathing was changed... Wouldn't that defeat the purpose of mass changes, or be extremely implementation specific?

desht replied:
Fair enough.  I wasn't overly convinced about including this one.

(The implementation needs to maintain a queue of pending relights for DEFERRED - it then processes those over several ticks, making sure not to spend too much time per tick doing the relighting).

Wolvereness commented on src/main/java/org/bukkit/block/MassBlockUpdate.java:
```diff
@@ -0,0 +1,140 @@
+     * @param level the brightness level, between 0 and 15 inclusive
+     * @throws IllegalStateException if the relighting strategy is not RelightingStrategy.STATIC
+     */
+    public void setLightLevel(int level);
```
I don't have all of the API memorized, but isn't this something we don't provide access to _anywhere_ else? If so, I already have nightmares of people making mass update objects just to change light level.

desht replied:
Nope, there's no way to change light levels in the API.  I've spent a bit of time researching this, and from what I can tell, there's no way of overriding light levels on a per-block basis.  Or rather you can set the light level of an individual block with a NMS method, but neither of the block update packets (0x34 or 0x35) allow for sending lighting changes to clients.  The only way to do it is in chunk update packets (0x33 or 0x38).

Your point about abuse of this API solely to change light levels is well made, though.

desht commented:
OK, after some more consideration, and in light of the criticism of attempt 1 at this, I hereby present attempt 2.  It's more or less a complete rewrite:
- Much less implementation-dependent, I hope.  A mass block updates is basically considered as a transaction: specify the blocks to be changed, and then "commit" the changes.
- Dropped the RelightingStrategy idea completely.  Default behaviour is now like the original DEFERRED, where block updates (not just relighting updates) are applied over one or more ticks, taking care not to spend too much time in each tick.
- Creation is now via getServer().createMassBlockUpdate() instead of via the World object; a MBU object is no longer tied to a single world, and can be used to make updates in multiple worlds.

Areas of potential controversy:
- I included a method to tune the maximum time per tick spent doing block updates, as a percent of one tick.  I'm not convinced that this should be included, although it might be a useful tunable.
- I included a method to statically set the lighting for all updated blocks (behaving like STATIC relighting strategy used to).  I know there are some reservations about this; it would definitely be useful, but I can see the possibility for abuse too.
- I've given two ways of specifying the blocks to be updated: 1) do a sequence of addBlockUpdate() calls, or 2) use setBlockUpdateProvider() to specify an Iterable<BlockUpdateRecord>.  The second method allows for the possibility of supplying updates on-demand, without needing to preallocate a big buffer of pending updates which could be a significant memory impact.
- Added a BlockUpdateRecord class, which is basically a record of (Location, material Id, data).  Right now its only constructors take a Location & Material, and possibly the data - if other overloaded constructors are needed, that's fine.
- The addBlockUpdate() method requires a BlockUpdateRecord which I admit feels a little clunky, but I'm not sure if there's a better way - I'm open to suggestions here.  Again, if other overloaded constructors are needed, that's fine.  Example usage is here: https://github.com/desht/buster2/blob/master/src/main/java/me/desht/buster2/buster2.java#L175

me4502 commented:
It would be nice to have a way to do it by ID, not just using Materials.

Wolvereness commented:
@me4502 We really should never be using ID.

Wolvereness commented:
I do think the BlockUpdateRecord wrapper would be counter-intuitive for what's being accomplished here. For mass edits, having a location object per edit stored (as part of our API) is bad. This also means the clone() call.

What'd I'd really like to see is the entire use-case for this type of API before more work is done. Obvious case is something like WorldEdit; what's being provided wont possibly fill that gap. If an API can't be designed to fit the purposes as well as be sane, then it's not worth having at all.

Less is more. Provide less API and use the implementation for dirty work. I still see things like lighting and 'styles'. Concepts like "update time" are going to be obtuse for everyone, both implementation and plugin dev. If you're doing an edit large enough to warrant splitting it, then you should either split it _manually_ or do it when the server is offline. Trying to get a 1-size fits all is not going to work; it really feels like it's designed to perfectly fill one use-case with little to no code on the plugin's side. Plugins being verbose and explicit is NOT a bad thing (as long as there can exist an efficient way). Remember, the JIT is there, and most operations are trivial.

The non-triviality part of the use-case really does render all these wrapper objects as counter intuitive.

me4502 commented:
@Wolvereness With CraftBook, I use WorldEdit's BlockID and ItemID classes, as they are kind of the abstract way of easily specifying an ID. Because WorldEdit is supported on multiple platforms, and I plan to add that support to CraftBook eventually.

desht commented:
@Wolvereness fair enough.  Is here an appropriate place for starting a use-case discussion?

These are some cases that immediately spring to mind:
- World editing plugins (WorldEdit being the prime example)
- Loading schematics into a world (may be included in the above case, but not necessarily)
- Minigame plugins where an arena or gameboard needs to be constructed/updated/reset regularly.  My own ChessCraft plugin falls into this category, hence my interest in the topic.
- "Apocalypse" plugins, e.g. creating floods, volcanoes, earthquakes etc., where a lot of terrain needs to be modified quickly.

I guess we're all clear on the basic problem that needs to be addressed; the current Bukkit block changing calls (i.e. the Block setXXX methods) are fine for changing one or two or a few blocks, but they don't scale at all well.

Wolvereness commented:
All but the last case would be indicative of some type of 'schematic' methodology in Bukkit. If you actually consider an "apocalypse" plugin, I think it'd be very clear that the protocol for minecraft is insufficient. I'd expect massive 'flooding, volcanoes, fissures' to all be in the category of "regenerate the chunk"; they're fundamentally terrain changes anyway.

desht commented:
OK, a generalised schematic interface in Bukkit, then.  What operations do we want for a schematic?
- Initialise an empty schematic
- Copy from region in world
- Paste to region in world
- Set/get block-id/data in the schematic
- Masking: the ability to define what blocks in the schematic actually get pasted
- Load from file (format defined by implementation; MCEdit would be an obvious choice)
- Save to file (format defined by implementation)

Re: masking.  Think about a hollow sphere made of stone.  It has a cubic boundary, but when pasting this into the world, you want the blocks inside the sphere's boundary to be replaced by air, but the blocks outside the sphere to be untouched.  So a means of stating "this block is not part of the schematic" is needed.

aadnk commented:
A system for loading and saving schematics seems interesting, but I still think you would need a mechanism for breaking up block changes into smaller steps and aggregating the lighting calculation to implement it, like MassBlockUpdate. Perhaps schematics could be implemented on top of that feature some time in the future?

@Wolvereness I understand your concern for constructing or cloning a location object for every edit, but I'm not convinced it would prove too costly. The JVM is incredibly efficient at allocating objects, and it may actually optimize away the defensive clone through [escape analysis](http://docs.oracle.com/javase/7/docs/technotes/guides/vm/performance-enhancements-7.html#escapeAnalysis), especially if the objects are short-lived and not stored in an array list. 

In fact, I think I would scrap the "addBlockUpdate" method for that very reason. You're better of allocating the update objects on-the-fly, as it may save memory and improve the chance of optimizing away a lot of copying (of course, this should be [benchmarked](http://stackoverflow.com/questions/16419233/best-practice-for-creating-millions-of-small-temporary-objects). Besides, an API consumer can always call setBlockUpdateProvider() with an array list if they want to replicate the addBlockUpdate behavior. The only difference is that multiple plugins can't easily share the same MassBlockUpdate, but I guess that was never the intention anyway. With that removed, you can also remove setUpdateBufferSize().

As for setMaxUpdateTimePerTick(), I agree with Wolvereness completely. It can only ever be a poor guess as plugins can't possibly know how long they ought to be processing. The correct duration is entirely dependent on the current server load, which will vary constantly every tick. Plugins would probably just leave it as is, or perhaps make it a configurable setting. Either that, or they might start doing something like [TCP congestion avoidance algorithm](http://en.wikipedia.org/wiki/TCP_congestion_avoidance_algorithm) and raise the update time until ticks start overlapping (can probably be detected using System.currentTimeMillis()). But all of this is needlessly complicated, and it could lead to some interesting and dangerous behavior if two plugins start fighting over "bandwidth". 

Instead, I think you should treat this as a scheduling problem similar to scheduling threads in an operating system. Each MassBlockUpdater could be treated as a virtual thread (not a OS thread) that gets scheduled according to some algorithm, which could be left as an implementation detail. CraftBukkit could use something simple like [round-robin](http://en.wikipedia.org/wiki/Round-robin_scheduling), but the interface might expose a "priority" - just like a thread priority - that other implementations can take into account. It wouldn't be very difficult to implement - you could process these mass block updates [here](https://github.com/Bukkit/mc-dev/blob/master/net/minecraft/server/MinecraftServer.java#L288) in the main game loop, instead of sleeping for a millisecond with Thread.sleep(1). 

Perhaps this could be implemented more abstractly - create a new type of scheduled task, the time sharing task:

``` java
public interface TimeSharingTask {
    /***
     * Execute the current task for no longer than the given duration.
     * @param maxDuration - maximum duration in milliseconds.
     */
    public void run(int maxDuration); 
        // This could also return a boolean, to indicate if a task is completed or not
}
```

This should be an _internal_ interface in CraftBukkit, and each MassBlockUpdater would register one such time sharing task. The main game loop would then execute each time sharing task for 1 millisecond until the circular queue of time sharing tasks is empty (tasks would remove themselves when they're done) OR at least 50 milliseconds have elapsed since the last tick. To ensure that tasks aren't completely starved, there might be a minimum number of time sharing tasks executed per tick. 

Perhaps, to ensure that each plugin gets a fair time slice, there could be a restriction of only one time sharing task per plugin. That is, the mass block updater would share the same time sharing task.

What do you guys think?

Wolvereness commented:
@aadnk Personal experience dictates you to be wrong about how efficient the JVM is. @amaranth could explain more on that subject and we have quite a bit of data on the AABB pool to prove it's extreme effectiveness. Even the WorldEdit guys could tell you how miserably the Block wrapper fails for large modifications.

aadnk commented:
@Wolvereness Right, the AABB pool. I might be wrong, sure, but I would first want to benchmark a typical application for this API proposal. I've found that object pooling is redundant if your objects are very short lived, though it might be situational. I'll write a benchmark in Google Caliper tormorrow and see if it matters.

If it _is_ a problem, we could create some sort of object pool, or perhaps pass in a recycled BlockUpdateRecord in a custom iterator-like interface:

``` java
public interface BlockUpdateProvider {
    public boolean hasNext();

    public BlockUpdateRecord next(BlockUpdateRecord recycled);
}
```

Though, I admit that is a bit ugly.

Wolvereness commented:
@aadnk Creating a pool is, quite honestly, exacerbating the problem. Did you even consider what you are pooling? Look at it: https://github.com/desht/Bukkit/commit/0c996e0abd022e096b7dce6704c094803370414d#L2R101. It's cringe worthy in context.

aadnk commented:
@Wolvereness I mocked up a quick and dirty benchmark, comparing the following two methods (with [MockWorld](https://github.com/aadnk/BenchmarkBUR/blob/master/TestProject/src/com/comphenix/testing/MockWorld.java))

``` java
public void timeProvider(int reps) {
    for (int i = 0; i < reps; i++) {
        blockUpdate.setProvider(getProvider(16, 16, 16));
        blockUpdate.applyUpdates();
    }
}

public void timeDirect(int reps) {
    for (int i = 0; i < reps; i++) {
        MockWorld mock = blockUpdate.getWorld();

        // The direct approach
        for (int x = 0; x < 16; x++) {
            for (int y = 0; y < 16; y++) {
                for (int z = 0; z < 16; z++) {
                    mock.setBlock(x, y, z, (x * y * z) & 0xFF, 0);
                }
            }
        }
    }
}
```

Using the current BlockUpdateRecord, this yielded the following timing ([full code](https://github.com/aadnk/BenchmarkBUR/blob/1345c6d1dc231e22ca396c25cb1dbcceee89fc00/TestProject/src/com/comphenix/testing/Test.java))

```
Without the server flag:
benchmark     us linear runtime
 Provider 559.19 ==============================
   Direct   9.16 =

With the server flag:
benchmark     us linear runtime
 Provider 555.64 ==============================
   Direct   9.19 =
```

The BlockUpdateRecord method is over 60 times slower than the direct method. That's definitely not acceptable, but it is possible to speed this up significantly by storing the location as three int fields instead of the Location class ([full code](https://github.com/aadnk/BenchmarkBUR/commit/73ba0f69f3ec4d4f65000494642c877692d97884)):

```
No server flag:
benchmark    us linear runtime
 Provider 70.03 ==============================
   Direct  9.14 ===

Server flag:
benchmark    us linear runtime
 Provider 69.97 ==============================
   Direct  9.16 ===
```

Recycling the current BlockUpdateRecord by modifying the iterator cuts off another 30 us ([full code](https://github.com/aadnk/BenchmarkBUR/commit/f112be0560a1d3074a5ff1257722d657a651fd8f)):

```
No server flag:
benchmark    us linear runtime
 Provider 43.81 ==============================
   Direct  9.31 ======

Server flag:
benchmark    us linear runtime
 Provider 43.40 ==============================
   Direct  9.17 ======
```

Switching to C#-style iterators shaved off another 3 us ([full code](https://github.com/aadnk/BenchmarkBUR/commit/d9558b18012eefaaa2dcd7850e9dc41882cb93a0)), but I haven't been able to reduce it any further:

```
Not server:
benchmark    us linear runtime
 Provider 40.46 ==============================
   Direct  9.20 ======

Server:
benchmark    us linear runtime
 Provider 40.30 ==============================
   Direct  9.13 ======
```

That might be good enough. But it's clear that the fastest method would be something like the direct method. I suggest providing a MockWorld-like wrapper around the Chunk data, and passing that to a method that is aware of the timeout:

``` java
interface BlockUpdator {
    // TODO: Rename MockWorld to something more suitable
    public boolean update(MockWorld world, Timeout timeout);
}

// Typical implementation
public class TestBlockUpdator implements BlockUpdator {
    private int x = 0;

    @Override
    public boolean update(MockWorld world, Timeout timeout);
        for (; x < 1000; x++) {
            for (int y = 0; y < 127; y++) {
                world.setBlock(x, y, 0, 1, 0);
            }

            if ((x % 10) == 0 && timeout.hasTimedOut())
                return false;
        }
        // We are done
        return true;
    }
}

public static class Timeout {
    public Timeout(int delay) {
        this.timeout = System.currentTimeMillis() + delay;
    }
    /**
     * Determine if the current virtual thread has timed out.
     * @return TRUE if it has, FALSE otherwise.
     */
    public boolean hasTimedOut() {
        return timeout < System.currentTimeMillis();
    }
}
```

That should be close to the DIRECT method.

desht commented:
Thanks @aadnk - some very interesting research there.

Here's some more food for thought:
- Yes, a schematic-style interface would certainly be useful, but I don't think it's sufficient.  By schematic, I'm taking to mean a (cuboid?) template of blocks which could be pasted into a world somehow.  But that would negate the possibility of procedurally-generated updates (e.g. generating structures on the fly from some mathematical algorithm), which pretty much requires a BlockUpdateProvider-style interface.
- OK, a BlockUpdateRecord with an embedded Location isn't going to fly for a load of pre-generated records.  It might work for on-the-fly generation, especially with a mutable, recyclable object, but that's not really flexible enough.
- Using 3 integers instead of a Location is certainly more efficient, and a Location isn't really the appropriate class here anyway, given that it stores a World object, 3 doubles and 2 floats when all we care about is 3 integers (assuming that one mass update operation is tied to one world, which is probably a good idea).
- We have two proposed approaches here, which are not particularly compatible with each other: the template concept - a static array of block/data values to be applied at a given location in the world, and the block-update concept - an iterable stream of block update records (location/block/data).  Each has its place, IMHO, and the ideal would be an interface that can cope with both.
- How to efficiently apply the updates (plus lighting etc.) is an implementation detail that needs to be isolated entirely from the interface; plugins shouldn't know how it's done (although the interface may need to specify that the update could take multiple ticks).  If updates can't be applied _quickly_, they at least need to be applied _carefully_, so as not to impact overall server performance.

aadnk commented:
@desht This is a fairly basic feature for making mass-edit plugins perform well and be backwards-compatible (things like WorldEdit, LogBlock), which is used by a sizable margin of server owners out there. It's definitely in everyone's interest to make NMS access redundant. So, seeing that this discussion had sort of stagnated, I figured I'd give my take on it to maybe speed things up a bit.

I think the schematic style is basically a special case of the BlockUpdateProvider-style. Schematics would make it fairly easy for CraftBukkit to divide the work up in pieces, but with additional memory overhead. However, I don't think it's necessary to go down the schematics route at all. If we use the correct abstraction for the BlockUpdateProvider that avoids too many object allocations, we can get near the same speed for less memory allocations and more power.

In that vein, I've been working on a TimeSharingTask implementation in CraftBukkit, and I think I've come up with a fairly good abstraction. Each "runnable" task will be called [FiberTask](https://github.com/aadnk/BenchmarkBUR/blob/fibers/TestProject/src/com/comphenix/testing/fibers/FiberTask.java) and be submitted to a [FiberExecutor](https://github.com/aadnk/BenchmarkBUR/blob/fibers/TestProject/src/com/comphenix/testing/fibers/FiberExecutor.java) for execution on the main thread. I use [round-robin](https://github.com/aadnk/BenchmarkBUR/blob/fibers/TestProject/src/com/comphenix/testing/fibers/RoundRobinExecutor.java) for testing, and there doesn't seem to be a lot of overhead ([test code](https://github.com/aadnk/BenchmarkBUR/blob/fibers/TestProject/src/com/comphenix/testing/Test.java)):

```
88% Scenario{vm=java, trial=0, benchmark=Direct, timeSlice=16} 823333.54 ns; ?=6593.15 ns @ 3 trials

  benchmark timeSlice  us linear runtime
  benchmark timeSlice   ms linear runtime
UpdateWorld         1 2.72 ==============================
UpdateWorld         2 2.66 =============================
UpdateWorld         8 2.61 ============================
UpdateWorld        16 2.63 =============================
     Direct         1 2.62 ============================
     Direct         2 2.60 ============================
     Direct         8 2.61 ============================
     Direct        16 2.62 ============================

vm: java
trial: 0
```

Now, all of these interfaces will be internal in CraftBukkit, and used by CraftMassBlockUpdate. Writing a FiberTask is certainly doable, but you have to strike a balance in checking ExecutionPermit often enough to catch it when it expires (and when you can exit the loop safely and still recover), while not calling it often enough to be a performance issue. Perhaps it's better for MassBlockUpdate to take care of this? I'd suggest a "buffer"-inspired interface to be included in Bukkit:

``` java
public interface BlockUpdater {
    /**
     * Update at most the given number of blocks in the current world.
     * @param world - current world.
     * @param maxBlockCount - maximum number of blocks to update.
     * @return TRUE if we are done updating blocks, FALSE otherwise.
     */
    // Again, MockWorld will be given a more suitable name, like DirectBlockAccess or FastWorldAccess
    public boolean update(MockWorld world, int maxBlockCount);
}
```

Then MassBlockUpdate will submit a FiberTask, and then take care of all the timing issues by gradually increasing the maxBlockCount until exceeding a certain threshold, like 200 microseconds per call. Then it's a simple matter of invoking update() until the FiberTask's execution permit has expired. 

Maybe you can get away with not having a maxBlockCount at all, and instead always update a single block. Still, certain implementations of BlockUpdater may benefit from aggregating multiple calls of update into one, and perhaps be simpler to write. We could also use an abstract class instead, with a default implementation of update:

``` java
public abstract class BlockUpdater {
    /**
     * Update at most the given number of blocks in the current world.
     * @param world - current world.
     * @param maxBlockCount - maximum number of blocks to update.
     * @return TRUE if we are done updating blocks, FALSE otherwise.
     */
    public boolean update(MockWorld world, int maxBlockCount) {
        for (int i = 0; i < maxBlockCount; i++) {
            if (updateBlock(world))
                return true;
        }
        return false;
    }

    /**
     * Update a single block in the current world.
     * @param world - current world.
     * @return TRUE if we are done updating blocks, FALSE otherwise.
     */
    public abstract boolean updateBlock(MockWorld world);
}
```

Keep in mind that the fiber task executor will be executed [here](https://github.com/Bukkit/mc-dev/blob/master/net/minecraft/server/MinecraftServer.java#L288) (and perhaps [here](https://github.com/Bukkit/mc-dev/blob/master/net/minecraft/server/MinecraftServer.java#L285) for a smaller duration), so they should never hold up regular processing. I don't know how long chunk.refreshChunk() will be executed however, but you can probably break that final portion up by only refreshing a certain number of chunks.

Comments?

orgin commented:
So is this going anywhere? The slow world.getBlockAt(..) / block.setType..() combo is just killing me!

turt2live commented:
@desht If you, or anyone else, is still interested in this PR, please address the following concerns:
- Code appears outdated and therefore not mergable
- A response to the large collection of performance tests above.

Once updated, this can be reviewed further. Thanks!

desht commented:
@turt2live I think this one's better off closed.  It won't get through in its current form; as @Wolvereness pointed out, this should really be done in terms of a schematic API with implementation details better isolated from the interface.

I may open a new PR at some point, but time constraints prevent me right now...
